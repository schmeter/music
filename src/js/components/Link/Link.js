import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';


const isExternal = (path) => /^https?:\/\/.+/.test(path);

class Link extends React.Component {
    render() {
        const { className, to, onClick, children } = this.props;
        const props = {
            className,
            onClick
        };
        const content = children || to;
        return isExternal(to) ? (
            <a
                target="_blank"
                href={to}
                {...props}
            >
                {content}
            </a>
        ) : (
            <NavLink
                exact
                activeClassName="active"
                to={to}
                {...props}
            >
                {content}
            </NavLink>
        );
    }
}

Link.propTypes = {
    to: PropTypes.string.isRequired,
    children: PropTypes.node,
    onClick: PropTypes.func,
    className: PropTypes.string
};

export default Link;
