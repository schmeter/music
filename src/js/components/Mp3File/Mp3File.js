import React, { useCallback } from 'react';
import PropTypes from 'prop-types';

import File from '../File';

const Mp3File = ({
    file,
    tracks,
    activeIndex,
    setActiveIndex,
    togglePlay,
}) => {
    const handleClickFile = useCallback(() => {
        const index = tracks.indexOf(file);

        if (index === activeIndex) {
            togglePlay();
        } else {
            setActiveIndex(index);
        }
    }, [
        file,
        tracks,
        activeIndex,
        setActiveIndex,
        togglePlay,
    ]);

    return (
        <File
            path={file.path}
            onClickFile={handleClickFile}
        >
            {file.tag.title}
        </File>
    );
};

Mp3File.propTypes = {
    file: PropTypes.object.isRequired,
    tracks: PropTypes.array.isRequired,
    activeIndex: PropTypes.number.isRequired,
    setActiveIndex: PropTypes.func.isRequired,
    togglePlay: PropTypes.func.isRequired,
};

export default Mp3File;
