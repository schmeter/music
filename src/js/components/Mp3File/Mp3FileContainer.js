import { connect } from 'react-redux';

import Mp3File from './Mp3File';
import { setActiveIndexAction, togglePlayAction } from '../../state/actions/audio';

export const mapDispatchToProps = (dispatch) => ({
    setActiveIndex: (activeIndex) => dispatch(setActiveIndexAction(activeIndex)),
    togglePlay: () => dispatch(togglePlayAction())
});

export const mapStateToProps = (state) => ({
    tracks: state.audio.tracks,
    activeIndex: state.audio.activeIndex
});

export default connect(mapStateToProps, mapDispatchToProps)(Mp3File);
