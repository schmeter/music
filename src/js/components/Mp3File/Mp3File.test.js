import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Mp3File from './Mp3File';
import Link from '../Link';

describe('Mp3File', () => {
  const mockedEvent = {
    preventDefault: jest.fn(),
  };
  const setActiveIndex = jest.fn();
  const togglePlay = jest.fn();
  const tracks = [{
    path: 'test1',
    tag: {
      title: 'test1',
    },
  }, {
    path: 'test2',
    tag: {
      title: 'test2',
    },
  }];

  it('renders correctly and uses togglePlay on click', () => {
    const props = {
      tracks,
      file: tracks[0],
      activeIndex: 0,
      setActiveIndex,
      togglePlay,
    };
    const component = mount(
      <BrowserRouter>
        <Mp3File {...props} />
      </BrowserRouter>,
    );

    expect(toJson(component)).toMatchSnapshot();

    component.find(Link).simulate('click', mockedEvent);

    expect(togglePlay).toHaveBeenCalled();
  });

  it('renders correctly and uses setActiveIndex on click', () => {
    const props = {
      tracks,
      file: tracks[1],
      activeIndex: 0,
      setActiveIndex,
      togglePlay,
    };
    const component = mount(
      <BrowserRouter>
        <Mp3File {...props} />
      </BrowserRouter>,
    );

    expect(toJson(component)).toMatchSnapshot();

    component.find(Link).simulate('click', mockedEvent);

    expect(setActiveIndex).toHaveBeenCalled();
  });
});
