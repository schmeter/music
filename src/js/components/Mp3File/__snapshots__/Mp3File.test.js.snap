// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mp3File renders correctly and uses setActiveIndex on click 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Mp3File
      activeIndex={0}
      file={
        Object {
          "path": "test2",
          "tag": Object {
            "title": "test2",
          },
        }
      }
      setActiveIndex={[MockFunction]}
      togglePlay={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      tracks={
        Array [
          Object {
            "path": "test1",
            "tag": Object {
              "title": "test1",
            },
          },
          Object {
            "path": "test2",
            "tag": Object {
              "title": "test2",
            },
          },
        ]
      }
    >
      <File
        onClickFile={[Function]}
        path="test2"
      >
        <Link
          className="file"
          onClick={[Function]}
          to="test2"
        >
          <NavLink
            activeClassName="active"
            className="file"
            exact={true}
            onClick={[Function]}
            to="test2"
          >
            <Link
              aria-current={null}
              className="file"
              onClick={[Function]}
              to={
                Object {
                  "hash": "",
                  "pathname": "/test2",
                  "search": "",
                  "state": null,
                }
              }
            >
              <LinkAnchor
                aria-current={null}
                className="file"
                href="/test2"
                navigate={[Function]}
                onClick={[Function]}
              >
                <a
                  aria-current={null}
                  className="file"
                  href="/test2"
                  onClick={[Function]}
                >
                  test2
                </a>
              </LinkAnchor>
            </Link>
          </NavLink>
        </Link>
      </File>
    </Mp3File>
  </Router>
</BrowserRouter>
`;

exports[`Mp3File renders correctly and uses togglePlay on click 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Mp3File
      activeIndex={0}
      file={
        Object {
          "path": "test1",
          "tag": Object {
            "title": "test1",
          },
        }
      }
      setActiveIndex={[MockFunction]}
      togglePlay={[MockFunction]}
      tracks={
        Array [
          Object {
            "path": "test1",
            "tag": Object {
              "title": "test1",
            },
          },
          Object {
            "path": "test2",
            "tag": Object {
              "title": "test2",
            },
          },
        ]
      }
    >
      <File
        onClickFile={[Function]}
        path="test1"
      >
        <Link
          className="file"
          onClick={[Function]}
          to="test1"
        >
          <NavLink
            activeClassName="active"
            className="file"
            exact={true}
            onClick={[Function]}
            to="test1"
          >
            <Link
              aria-current={null}
              className="file"
              onClick={[Function]}
              to={
                Object {
                  "hash": "",
                  "pathname": "/test1",
                  "search": "",
                  "state": null,
                }
              }
            >
              <LinkAnchor
                aria-current={null}
                className="file"
                href="/test1"
                navigate={[Function]}
                onClick={[Function]}
              >
                <a
                  aria-current={null}
                  className="file"
                  href="/test1"
                  onClick={[Function]}
                >
                  test1
                </a>
              </LinkAnchor>
            </Link>
          </NavLink>
        </Link>
      </File>
    </Mp3File>
  </Router>
</BrowserRouter>
`;
